### PR TITLE
Ms/equivarinat matrices

### PR DESCRIPTION
**Pull request by MatthiasSachs**
All tests asserting the equivariance properties of  `EuclidianMatrix` pass with tol = 1E-15. 

The test `println(@test(all(test_fio(basis; warntype = false))))` (see line 41 in test/test_EquivariantMatrix.jl) does NOT pass...

### CODE REVIEW

**Comment by MatthiasSachs** [comment]
`test/test_EquivariantMatrix.jl` not yet added to test collection. Thus, all tests are passing for now...

**Comment by cortner** [comment]
thank you - I'll take a look sap.

**Comment by cortner** [comment]
I'm looking into the read/write issue now.

**Comment by cortner** [comment]
Regarding the name, I think EquivariantMatrix is too general - it doesn't encode the specific symmetry. How about `EuclideanMatrix`? This would then also generalize nicely to `EuclideanTensor` if/when we get to that.

**Comment by cortner** [comment]
Another thing I've noticed is that you've based it off a fairly old commit. You'll need to rebase it onto the latest `main` please.

**Comment by cortner** [comment]
You'll see there that `EuclideanVector` has changed a bit and it would be good to model `EuclideanMatrix` on that. I hope that should require minimal work.

**Comment by MatthiasSachs** [comment]
> You'll see there that `EuclideanVector` has changed a bit and it would be good to model `EuclideanMatrix` on that. I hope that should require minimal work.

This is mostly done. However, 
1) I am not sure if the implementation of `coco_init(::EuclideanMatrix{CT})` is correct?
2) Moreover, somehow there is an error thrown which relates to the conversion between SMatrix and EuclideanMatrix.

Would you mind taking a brief look at that (I think this might be much easier for you to resolve since you know the parts of the code well), or alternatively, we could meet at the beginning of next week?

**Comment by cortner** [comment]
Error message:
```
 [1] convert(T::Type, φ::EuclideanMatrix{Float64})
   @ ACE ~/gits/temp/ms.ACE.jl/src/properties.jl:28
 [2] push!(a::Vector{EuclideanMatrix{ComplexF64}}, item::EuclideanMatrix{Float64})
   @ Base ./array.jl:994
....
```

It happens in this line: 
```Julia
push!(vals, U[irow, icol])
```
The computed cc `U[irow, icol]` is of type `item::EuclideanMatrix{Float64}` and get's appended to a vector with eltype `EuclideanMatrix{ComplexF64}`. We need a `convert` implementation for this. THat's easy to do, but it concerns me. Why are we creating a complex-valued CC matrix but then fill it with real-valued CCs?

**Comment by MatthiasSachs** [comment]
> Error message:
> 
> ```
>  [1] convert(T::Type, φ::EuclideanMatrix{Float64})
>    @ ACE ~/gits/temp/ms.ACE.jl/src/properties.jl:28
>  [2] push!(a::Vector{EuclideanMatrix{ComplexF64}}, item::EuclideanMatrix{Float64})
>    @ Base ./array.jl:994
> ....
> ```
> 
> It happens in this line:
> 
> ```julia
> push!(vals, U[irow, icol])
> ```
> 
> The computed cc `U[irow, icol]` is of type `item::EuclideanMatrix{Float64}` and get's appended to a vector with eltype `EuclideanMatrix{ComplexF64}`. We need a `convert` implementation for this. THat's easy to do, but it concerns me. Why are we creating a complex-valued CC matrix but then fill it with real-valued CCs?

I fixed other parts of the code so that now there is no conversion required! It all seems to work now.

File: src/properties.jl [hunk 1]
```diff
@@ -246,39 +246,42 @@ rmatrices = Dict(
...
 m[$(φ.val[3,1]), $(φ.val[3,2]), $(φ.val[3,3])]")
 
-real(φ::EuclideanMatrix) = EuclideanMatrix(φ.val)
-complex(φ::EuclideanMatrix) = EuclideanMatrix(φ.val)
+real(φ::EuclideanMatrix) = EuclideanMatrix(real.(φ.val))
+complex(φ::EuclideanMatrix) = EuclideanMatrix(complex(φ.val))
 complex(::Type{EuclideanMatrix{T}}) where {T} = EuclideanMatrix{complex(T)}
 
 +(x::SMatrix{3}, y::EuclideanMatrix) = EuclideanMatrix(x + y.val)
 Base.convert(::Type{SMatrix{3, 3, T, 9}}, φ::EuclideanMatrix) where {T} =  convert(SMatrix{3, 3, T, 9}, φ.val)
```
Comment (cortner) [inline src/properties.jl:263]: where is this needed?

File: src/properties.jl [hunk 2]
```diff
@@ -246,39 +246,42 @@ rmatrices = Dict(
 
 #---------------------- Equivariant matrices
 
-struct EuclideanMatrix{T} <: AbstractProperty where T<:Real
-   val::SMatrix{3, 3, Complex{T}, 9}
+struct EuclideanMatrix{T} <: AbstractProperty
+   val::SMatrix{3, 3, T, 9}
 end
 
 Base.show(io::IO, φ::EuclideanMatrix) =
```
Comment (cortner) [inline src/properties.jl:253]: I'm not sure about this stype of printing; it suggests 3 vectors rather than one matrix. Would it be ok to print it similarly to the spherical matrix, but with an "e" label instead of a "y" label? (yes, I also don't like the "m" label)

File: src/properties.jl [hunk 3]
```diff
@@ -246,39 +246,42 @@ rmatrices = Dict(
...
 EuclideanMatrix(T::DataType=Float64) = EuclideanMatrix{T}()
 
 
-function filter(φ::EuclideanMatrix, grp::O3, b::Array)
-   if length(b) <= 1 #MS: Not sure if this should be here
+function filter(φ::EuclideanMatrix, grp::O3, bb::Array)
+   if length(bb) == 0  # no zero-correlations allowed 
+      return false 
+   end
+   if length(bb) == 1 #MS: Not sure if this should be here
```
Comment (cortner) [inline src/properties.jl:279]: I don't remember the issue - same with EuclideanVector. If codes run ok, leave it for now but file an issue to investigate this. Do you have an idea?

File: src/properties.jl [hunk 4]
```diff
@@ -304,16 +307,17 @@ end
 
 coco_init(phi::EuclideanMatrix{CT}, l, m, μ, T, A) where {CT<:Real} = (
       (l == 2 && abs(m) <= 2 && abs(μ) <= 2)
-         ? vec([EuclideanMatrix{CT}(conj.(transpose(mrmatrices[(m,μ,i,j)]))) for i=1:3 for j=1:3])
+         ? vec([EuclideanMatrix(conj.(transpose(mrmatrices[(m,μ,i,j)]))) for i=1:3 for j=1:3])
          : coco_zeros(phi, l, m, μ, T, A)  )
 
 
-coco_init(::EuclideanMatrix{CT}) where {CT<:Real} = [EuclideanMatrix{CT}(SMatrix{3,3,Float64,9}([1.0,0,0,0,1.0,0,0,0,1.0]))]
```
Comment (cortner) [inline src/properties.jl]: I think you want 
```
EuclideanMatrix{CT}(SMatrix{3,3,CT}([1.0,0,0,0,1.0,0,0,0,1.0]))
```
If I remember correctly this is the implementation for zero-correlations. So you are right that identity is the only zero-correlation there is, but you probably need to specify the right eltype.

File: src/properties.jl [hunk 5]
```diff
@@ -304,16 +307,17 @@ end
...
 coco_init(phi::EuclideanMatrix{CT}, l, m, μ, T, A) where {CT<:Real} = (
       (l == 2 && abs(m) <= 2 && abs(μ) <= 2)
-         ? vec([EuclideanMatrix{CT}(conj.(transpose(mrmatrices[(m,μ,i,j)]))) for i=1:3 for j=1:3])
+         ? vec([EuclideanMatrix(conj.(transpose(mrmatrices[(m,μ,i,j)]))) for i=1:3 for j=1:3])
          : coco_zeros(phi, l, m, μ, T, A)  )
 
 
-coco_init(::EuclideanMatrix{CT}) where {CT<:Real} = [EuclideanMatrix{CT}(SMatrix{3,3,Float64,9}([1.0,0,0,0,1.0,0,0,0,1.0]))]       
+#coco_init(::EuclideanMatrix{CT}) where {CT<:Real} = [EuclideanMatrix(SMatrix{3,3,Complex{CT},9}([1.0,0,0,0,1.0,0,0,0,1.0]))]       
 coco_type(φ::EuclideanMatrix) = typeof(complex(φ))
```
Comment (cortner) [inline src/properties.jl:315]: so this confuses me now. In the previous commit the CCs there were produced were all real. But here you are saying they should be complex. So which is it?

File: src/properties.jl [hunk 6]
```diff
@@ -304,16 +307,17 @@ end
...
 coco_type(::Type{EuclideanMatrix{T}}) where {T} = EuclideanMatrix{complex(T)}
 
 # This is slightly different from implementation in EuclideanVector!
-coco_zeros(::EuclideanMatrix, ll, mm, kk, T, A) =  EuclideanMatrix.(zeros(SMatrix{3, 3, Complex{T}, 9},9))
+coco_zeros(φ::EuclideanMatrix, ll, mm, kk, T, A) =  zeros(typeof(complex(φ)), 9)
```
Comment (cortner) [inline src/properties.jl:319]: same here, why are you enforcing complex CCs?

**Review by cortner — CHANGES_REQUESTED** [review]
looks mostly ok, just a few more smaller things to address please. Can you then also add the new tests to the test suite please? 

The zero-correlation implementation looks roughly ok, but I added a small comment.

File: src/properties.jl (hunk 2 above)
Comment (MatthiasSachs) [reply src/properties.jl:253]: Done! But I am wondering whether we should have a different label than the label of Euclidean vectors (since EuclideanMatrix satisfies a different equivariance property)

File: src/properties.jl (hunk 5 above)
Comment (MatthiasSachs) [reply src/properties.jl:315]: This is all very confusing! But I know for sure that the coupling coefficients should be complex (same as in the implementation of `EuclideanVector`). Thus, my understanding is that both the `coco_init` function as well as the `coco_zeros` function should return complex valued matrices.

File: src/properties.jl (hunk 1 above)
Comment (MatthiasSachs) [reply src/properties.jl:263]: not sure, I kind of copied the implementations of `EuclideanVector` here... Should I check, or should we introduce a new abstract type to avoid code redundancy?

File: src/properties.jl (hunk 3 above)
Comment (MatthiasSachs) [reply src/properties.jl:279]: The code runs. Issues is filed!

File: src/properties.jl (hunk 4 above)
Comment (MatthiasSachs) [reply src/properties.jl]: included in issue #100

File: src/properties.jl (hunk 6 above)
Comment (MatthiasSachs) [reply src/properties.jl:319]: s.a.

File: src/properties.jl (hunk 2 above)
Comment (cortner) [reply src/properties.jl]: I think it can be clear from the context that it is a matrix. If it were e and then three vectors then I would be confused.

File: src/properties.jl (hunk 1 above)
Comment (cortner) [reply src/properties.jl:267]: ok, leave it for now.

**Comment by MatthiasSachs** [comment]
sorry, just pushed a fix for a small bug which I noticed today in the morning.

**Comment by MatthiasSachs** [comment]
@cortner I think this PR is ready to merge. Do you approve?

**Comment by cortner** [comment]
I will take a look. We will also need to merge in the latest main branch, or rebase.

**Comment by MatthiasSachs** [comment]
@cortner: I just merged the upstream/main into this branch.

**Comment by MatthiasSachs** [comment]
Argghh... it's not obvious to me how to fix this error!

**Comment by cortner** [comment]
I wil try - no promises about time-scale

**Comment by cortner** [comment]
what error? I only get a warning?

**Comment by MatthiasSachs** [comment]
This one:
```julia
ERROR: LoadError: MethodError: namedtuple(::Tuple{}) is ambiguous. Candidates:
  namedtuple(names::Tuple{Vararg{Symbol, N}}) where N in NamedTupleTools at /Users/msachs2/.julia/packages/NamedTupleTools/GZxRn/src/NamedTupleTools.jl:196
  namedtuple(names::Tuple{Vararg{String, N}}) where N in NamedTupleTools at /Users/msachs2/.julia/packages/NamedTupleTools/GZxRn/src/NamedTupleTools.jl:198
Possible fix, define
  namedtuple(::Tuple{})
Stacktrace:
  [1] select(nt::NamedTuple{(:be, :n, :l, :m), Tuple{Symbol, Int64, Int64, Int64}}, ks::Tuple{})
    @ NamedTupleTools ~/.julia/packages/NamedTupleTools/GZxRn/src/NamedTupleTools.jl:270
  [2] _broadcast_getindex_evalf
    @ ./broadcast.jl:648 [inlined]
  [3] _broadcast_getindex
    @ ./broadcast.jl:621 [inlined]
  [4] getindex
    @ ./broadcast.jl:575 [inlined]
  [5] copy
    @ ./broadcast.jl:922 [inlined]
  [6] materialize
    @ ./broadcast.jl:883 [inlined]
  [7] _sparsify_component!(basis1p::ACE.EllipsoidBondEnvelope{Float64}, keep::Vector{NamedTuple{(:be, :n, :l, :m), Tuple{Symbol, Int64, Int64, Int64}}})
    @ ACE ~/.julia/dev/ACE/src/product_1pbasis.jl:297
  [8] sparsify!(basis1p::ACE.Product1pBasis{4, Tuple{Categorical1pBasis{:be, :be, 2, Symbol}, ACE.Rn1pBasis{Float64, PolyTransform{Int64, Float64}, ACE.OrthPolys.OrthPolyBasis{Float64}, :rr, :n, ACE.DState{NamedTuple{(:rr,), Tuple{SVector{3, Float64}}}}}, ACE.Ylm1pBasis{Float64, :rr, :l, :m, ACE.DState{NamedTuple{(:rr,), Tuple{SVector{3, ComplexF64}}}}}, ACE.EllipsoidBondEnvelope{Float64}}, ComplexF64}, keep::Vector{NamedTuple})
    @ ACE ~/.julia/dev/ACE/src/product_1pbasis.jl:277
  [9] clean_1pbasis!(basis::PIBasis{ACE.Product1pBasis{4, Tuple{Categorical1pBasis{:be, :be, 2, Symbol}, ACE.Rn1pBasis{Float64, PolyTransform{Int64, Float64}, ACE.OrthPolys.OrthPolyBasis{Float64}, :rr, :n, ACE.DState{NamedTuple{(:rr,), Tuple{SVector{3, Float64}}}}}, ACE.Ylm1pBasis{Float64, :rr, :l, :m, ACE.DState{NamedTuple{(:rr,), Tuple{SVector{3, ComplexF64}}}}}, ACE.EllipsoidBondEnvelope{Float64}}, ComplexF64}, typeof(real), Float64, ComplexF64})
    @ ACE ~/.julia/dev/ACE/src/pibasis.jl:221
 [10] clean_pibasis!(basis::ACE.SymmetricBasis{PIBasis{ACE.Product1pBasis{4, Tuple{Categorical1pBasis{:be, :be, 2, Symbol}, ACE.Rn1pBasis{Float64, PolyTransform{Int64, Float64}, ACE.OrthPolys.OrthPolyBasis{Float64}, :rr, :n, ACE.DState{NamedTuple{(:rr,), Tuple{SVector{3, Float64}}}}}, ACE.Ylm1pBasis{Float64, :rr, :l, :m, ACE.DState{NamedTuple{(:rr,), Tuple{SVector{3, ComplexF64}}}}}, ACE.EllipsoidBondEnvelope{Float64}}, ComplexF64}, typeof(real), Float64, ComplexF64}, ACE.Invariant{Float64}, ACE.O3{:l, :m}, typeof(real), ACE.Invariant{Float64}}; atol::Float64)
    @ ACE ~/.julia/dev/ACE/src/symmbasis.jl:250
 [11] clean_pibasis!(basis::ACE.SymmetricBasis{PIBasis{ACE.Product1pBasis{4, Tuple{Categorical1pBasis{:be, :be, 2, Symbol}, ACE.Rn1pBasis{Float64, PolyTransform{Int64, Float64}, ACE.OrthPolys.OrthPolyBasis{Float64}, :rr, :n, ACE.DState{NamedTuple{(:rr,), Tuple{SVector{3, Float64}}}}}, ACE.Ylm1pBasis{Float64, :rr, :l, :m, ACE.DState{NamedTuple{(:rr,), Tuple{SVector{3, ComplexF64}}}}}, ACE.EllipsoidBondEnvelope{Float64}}, ComplexF64}, typeof(real), Float64, ComplexF64}, ACE.Invariant{Float64}, ACE.O3{:l, :m}, typeof(real), ACE.Invariant{Float64}})
    @ ACE ~/.julia/dev/ACE/src/symmbasis.jl:243
 [12] ACE.SymmetricBasis(φ::ACE.Invariant{Float64}, symgrp::ACE.O3{:l, :m}, pibasis::PIBasis{ACE.Product1pBasis{4, Tuple{Categorical1pBasis{:be, :be, 2, Symbol}, ACE.Rn1pBasis{Float64, PolyTransform{Int64, Float64}, ACE.OrthPolys.OrthPolyBasis{Float64}, :rr, :n, ACE.DState{NamedTuple{(:rr,), Tuple{SVector{3, Float64}}}}}, ACE.Ylm1pBasis{Float64, :rr, :l, :m, ACE.DState{NamedTuple{(:rr,), Tuple{SVector{3, ComplexF64}}}}}, ACE.EllipsoidBondEnvelope{Float64}}, ComplexF64}, typeof(real), Float64, ComplexF64}, _real::Function)
    @ ACE ~/.julia/dev/ACE/src/symmbasis.jl:176
 [13] #SymmetricBasis#313
    @ ~/.julia/dev/ACE/src/symmbasis.jl:103 [inlined]
 [14] ACE.SymmetricBasis(φ::ACE.Invariant{Float64}, basis1p::ACE.Product1pBasis{4, Tuple{Categorical1pBasis{:be, :be, 2, Symbol}, ACE.Rn1pBasis{Float64, PolyTransform{Int64, Float64}, ACE.OrthPolys.OrthPolyBasis{Float64}, :rr, :n, ACE.DState{NamedTuple{(:rr,), Tuple{SVector{3, Float64}}}}}, ACE.Ylm1pBasis{Float64, :rr, :l, :m, ACE.DState{NamedTuple{(:rr,), Tuple{SVector{3, ComplexF64}}}}}, ACE.EllipsoidBondEnvelope{Float64}}, ComplexF64}, symgrp::ACE.O3{:l, :m}, Bsel::ACE.CategorySparseBasis; isreal::Bool, kwargs::Base.Iterators.Pairs{Symbol, ACE.Utils.var"#5#7", Tuple{Symbol}, NamedTuple{(:filterfun,), Tuple{ACE.Utils.var"#5#7"}}})
    @ ACE ~/.julia/dev/ACE/src/symmbasis.jl:96
 [15] #SymmetricBasis#310
    @ ~/.julia/dev/ACE/src/symmbasis.jl:85 [inlined]
 [16] SymmetricBond_basis(ϕ::ACE.Invariant{Float64}, env::ACE.EllipsoidBondEnvelope{Float64}, Bsel::SparseBasis; RnYlm::Nothing, bondsymmetry::Nothing, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ ACE.Utils ~/.julia/dev/ACE/src/utils.jl:97
 [17] SymmetricBond_basis(ϕ::ACE.Invariant{Float64}, env::ACE.EllipsoidBondEnvelope{Float64}, Bsel::SparseBasis)
    @ ACE.Utils ~/.julia/dev/ACE/src/utils.jl:78
 [18] top-level scope
    @ ~/.julia/dev/ACE/test/test_bondbasisselectors.jl:38
in expression starting at /Users/msachs2/.julia/dev/ACE/test/test_bondbasisselectors.jl:36
```

**Comment by cortner** [comment]
ok, I can reproduce

**Comment by cortner** [comment]
Ok - the problem is here: 
```
env = ACE.EllipsoidBondEnvelope(r0cut, rcut, zcut;floppy=false, λ= .5)
property = Invariant() ... 
basis = ACE.Utils.SymmetricBond_basis(property, env, Bsel; )
```
you are trying to create a symmetric basis but it isn't a basis, so the basis spec contains only empty tuples. That's a weird edge case and I'm not even sure I want to allow this... Is this every going to occur outside of a test?

**Comment by cortner** [comment]
See here : once we sparsity (in this case just "clean up") ... 
```
function _sparsify_component!(basis1p, keep)
   # get rid of all info we don't need 
   syms = symbols(basis1p)
   keep1 = unique( select.(keep, Ref(syms)) )
   .... 
```
I can't "select" a subset of the basis specs because there is no spec. It is at odds with the entire design of the basis framework.

**Comment by MatthiasSachs** [comment]
> Ok - the problem is here:
> 
> ```
> env = ACE.EllipsoidBondEnvelope(r0cut, rcut, zcut;floppy=false, λ= .5)
> property = Invariant() ... 
> basis = ACE.Utils.SymmetricBond_basis(property, env, Bsel; )
> ```
> 
> you are trying to create a symmetric basis but it isn't a basis, so the basis spec contains only empty tuples. That's a weird edge case and I'm not even sure I want to allow this... Is this every going to occur outside of a test?

I can see in the error message that the spec is empty but I don't yet understand why this is the case? I create a special basis selector within the call of `SymmetricBond_basis` and then call a constructor for `SymmetricBasis` using this basis selector. Why is what I am instantiating not a basis? Is the bond envelope `env` and the fact that we don't associate a symbol with that causing problems here? So far this is the standard way for me to create bond bases...

**Comment by cortner** [comment]
Any thoughts on this? Maybe there is a way to just change the tests? Or do you need me to figure out a way around this?
